specs/sharding/beacon-chain.md

```diff
@@ -179,19 +216,24 @@ class BeaconBlockBody(merge.BeaconBlockBody):  # [extends The Merge block body]
 
 ```python
 class BeaconState(merge.BeaconState):
-    # [Updated fields] (Warning: this changes with Altair, Sharding will rebase to use participation-flags)
-    previous_epoch_attestations: List[PendingAttestation, MAX_ATTESTATIONS * SLOTS_PER_EPOCH]
-    current_epoch_attestations: List[PendingAttestation, MAX_ATTESTATIONS * SLOTS_PER_EPOCH]
-    # [New fields]
+    # Blob builder registry.
+    blob_builders: List[Builder, BLOB_BUILDER_REGISTRY_LIMIT]
+    blob_builder_balances: List[Gwei, BLOB_BUILDER_REGISTRY_LIMIT]
     # A ring buffer of the latest slots, with information per active shard.
     shard_buffer: Vector[List[ShardWork, MAX_SHARDS], SHARD_STATE_MEMORY_SLOTS]
-    shard_gasprice: uint64
+    shard_sample_price: uint64
 ```
 
 ## New containers
 
-The shard data itself is network-layer only, and can be found in the [P2P specification](./p2p-interface.md).
-The beacon chain registers just the commitments of the shard data.
+### `Builder`
+
+```python
+class Builder(Container):
+    pubkey: BLSPubkey
+    # TODO: fields for either an expiry mechanism (refunding execution account with remaining balance) 
```

Comment:
One idea to avoid builder withdrawal transactions (which come with signing and inclusion overhead) is to put an expiry date on an account, to clean up passively. Either a fixed epoch when the builder gets withdrawn to execution state, or we can implement some inactivity check (withdraw funds if not used for N epochs, maybe specified by the builder).

That said, for initial sharding prototypes we don't need deposits or withdrawals, we can just pre-mine the builders into the testnet genesis state. Implementing a builder lifecycle (via EVM, via beacon state, or some buffer kind of thing) can go into a later PR.

Reply:
yeah for sure. don't need to solve the nitty gritty yet